<a href="https://colab.research.google.com/github/eriksh26/Calculator-JS/blob/main/Notebooks/1_1_Falling_penny/1_1_3_Falling_Penny.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 1.1.3: The Falling Penny (Modeling with Computation)

<br>



---



*Modeling and Simulation in Python*

Copyright 2021 Allen Downey, (License: [Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International](https://creativecommons.org/licenses/by-nc-sa/4.0/))

Revised, Mike Augspurger (2021-present)

<br>

---

## Modeling the Velocity of a Penny Falling

Now we're going to create a model to determine the final speed and time required for the penny to fall.

<br>

The statistician George Box famously said "All models are wrong, but
some are useful." The puzzle that modelers need to solve is how to create a model that combines accuracy (usefulness) and inaccuracy (efficiency) in the right mix: we want a model that gives us a usable answer without wasting time and resources.

<br>

<center>
<img src = https://github.com/MAugspurger/ModSimPy_MAugs/raw/main/Images_and_Data/Images/1_1/Wrong.PNG width = 400>
</center>


<br>

Generally, we start simple and add complexity as necessary.  The model we'll start with is a simple one: in fact, it will look a fair bit like a homework problem from a physics class.  This is an important connection to make, because most of your homework problems in physics or chemistry or biology *are* models--that is, they are simplications of a real world situation designed to help us understand the real world situation better.  You've done more modeling than you even knew!

### Investigation and Abstraction

In your investigation in the exercises of the last notebook (i.e. observing the *Mythbusters* video), you probably decided that drag was an important factor for the velocity of the penny, and you were correct.   But sometimes we implement a model that we know is too simple just to make sure our implementation is successful, and then we iterate on that model to add more complexity.

<br>

We'll do that here.  We're going to reduce our list of possible factors affecting the final velocity of the penny to two factors: gravitational force and height.  We'll assume the effect of all other factors is *negligible* (that is, they have only small effects), and so we will ignore them.  Engineers do this all the time, no matter how much the mathematicians resent it 😏

<br>

So we have completed the first half of abstraction: the selection of our key factors.  The second part of abstraction is defining the *relationships* between the key factors.  In the last notebook, we dropped a penny from a ladder (in our heads, anyway) and hypothesized that the relationship between time and velocity of a falling penny was linear:

<br>

$$v=Ct$$

<br>

Another form of "investigation" would be to look at the work of previous scientists.  A quick look at an intro physics book would reveal this work, in the form of the *kinematic equations*.  These equations tell us about the relationships between time, height and velocity for an object in free-fall (when drag is negligible).  

<br>

Happily, these equations tell us that our ladder investigation was valid!  The kinematic equation relating time and velocity in free-fall is:

$$v = g t$$

This model tells us that if the initial velocity is 0, the velocity after time $t$ is a function of the gravitational constant $g$, which at sea level is approximately $g=9.81$ $m/s^2$.

<br>

A second kinematic equation relates time and distance fallen $x$:

$$x = g t^2 / 2$$

<br>

With a little math magic, we can combine these two equations to find the relationship between our two factors and the final velocity of the penny:

<br>

$$ v = \sqrt{2gx}$$

<br>

This equation is our abstracted model for the final velocity of a penny-dropped from a height $x$ when drag is negligible.


🟨 🟨 Active Reading: Why might it make sense to ignore drag in our model, despite the fact that we know drag is important?

✅ ✅ The valaue of the drag is so small that it's negligible.



### Implementation and Validation

"Implementation" in this case is pretty straight-forward: we can just take the values from our particular situation and plug them into our model equation.

<br>

We could do this with a calculator, but let's use Python. First we'll create a variable to represent acceleration due to gravity in meters per second squared (m/s$^2$).

In [1]:
g = 9.8

A *variable* is a name that corresponds to a value. In this example, the name is `g` and the value is the number `9.8`.  When you run the above cell (click the `play` button or press `shift-enter`), nothing seems to happen.  But in fact something does: the code *assigns* the value `9.8` to the variable `g`.

<br>

Suppose we let the penny drop for $3.4$ seconds (s).  I'll create a variable to represent this time:

In [6]:
t = 3.4 #Units: seconds
print("The time is", t, "seconds.")

The time is 3.4


Notice that when I add the second line `t`, running the cell produces an output.  This doesn't change the variable at all: it just tells the computer to display the value of `t`.

<br>

Now we can compute the velocity of the penny after `t` seconds, and assign the answer to a variable named `v`:

In [4]:
v = g * t #Units: m/s
print("The speed is", v, "m/s.")

The speed is 33.32 m/s.


Python uses the symbol `*` for multiplication.  The other arithmetic operators are `+` for addition, `-` for subtraction, `/` for division, and `**` for exponentiation (like $5^2$).  The variable `v` now has been assigned a value.  We can display its value like this:

In [ ]:
v

After $3.4$ s, the velocity of the penny is about $33$ m/s (ignoring air resistance).  Now let's see how far it would travel during that time.  Remember from above that our equation for distance travelled is $x = g t^2 / 2$:

In [7]:
x = g * ((t**2)/ 2)
print("The position is:", x, "meters.")

The position is: 56.644 meters.


It would travel about $56$ m.  Now, going in the other direction, let's compute the time it takes to fall 381 m, the height of the Empire State Building.

In [14]:
h = 381

For this computation, we need the square root function, `sqrt`, which is provided by a *library* called NumPy, which contains functions related to the manipulation of numbers.  NumPy is built into Colab.  But before we can use it in our notebook,we have to *import* it.  Importing a library or function means we make it available to use, even though the code for that function is not in this particular notebook:

In [19]:
import numpy as np

Again using the kinematic equation that we used above, we can write:

In [20]:
t = np.sqrt(2 * h / g)
print(t)

8.817885349720552


With no air resistance, it would take about $8.8$ s for the penny to reach the sidewalk.

In [21]:
v = g * t
v

86.41527642726142

And its velocity on impact would be about $86$ m/s, or about 190 miles per hour.  That sounds like it could hurt.  We don't need to do much here in terms of validation, though.  Our intuition (aided by the *Mythbusters*) tells us that a penny travelling at 190 mph would quickly be slowed down by drag.  

<br>

This model needs some iteration....

### Iteration

We are now ready to enter the implementation/ validation/ iteration feedback loop!  

<br>

As the *Mythbusters* guys showed, ignoring air resistance leads to major errors. So we want to iterate our model: that is, we want to add in some new complication that will make the model more troublesome, but happily more accurate.  It turns out that once
the penny gets to about 29 m/s, the upward force of air resistance
equals the downward force of gravity, so the penny stops accelerating.
This is the *terminal velocity* of the penny in air.

<br>

Once the penny reaches terminal velocity, it doesn't matter how much farther it falls; it hits the sidewalk at about 29 m/s (65 mph).
That's much less than 86 m/s, as the simple model predicts.

<br>  In the exercises below, you'll be asked to perform this iteration.  Before we do that, here are some important notes about using variables in Python

<br>

---

## Summary and Exercises

As a first example, we modeled a penny dropped from the Empire State building, including gravity but ignoring air resistance.  In the exercises, you'll iterate the model a couple times in an attempt to improve its accuracy.

<br>

---

🟨 🟨

### Exercise 1

In this chapter we used the `sqrt` function from the NumPy library.  NumPy also provides a variable named `pi` that contains an approximation of the mathematical constant $\pi$.
We can import it like this:

In [23]:
from numpy import pi
pi

3.141592653589793

NumPy provides other functions we'll use, including `log`, `exp`,`square`,`power`, `sin`, and `cos`.  To find out how to use these functions, see https://numpy.org/doc/stable/reference/routines.math.html

<br>

Import `sin` and `cos` from NumPy and compute

<br>

$$sin^2 (\pi/4) + cos^2 (\pi/4)$$

<br>Note: A mathematical identity tells us that the answer should be $1$.  There are are multiple ways to do this, including using `power` or `square` from the NumPy library (remember to import them).  Try to find a couple ways to get the correct answer.  

<br>

IMPORTANT: work one step at a time.  First, make sure you get the correct result for $\pi/4$.  Then try $sin~ \pi/4$ (which should be about 0.707), etc...  Note that $sin^2 (\pi/4)$ is the same as $(sin (\pi/4))^2$, and should have a value of about 0.500.

In [31]:
from numpy import sin
from numpy import cos
v = pi/4
s = sin(v)
c = cos(v)
a = (sin(v))**2
b = (cos(v))**2
x = a + b
print(x)



1.0


Did you notice that you often do not get exactly $1$?  This is the difference between *analytical math* (which tells us that the answer is exactly $1$) and *computational math* (which routinely has can have (usually very small) rounding errors).


---
<br>

🟨 🟨

### Exercise 2

Suppose instead of dropping the penny, you throw it downward at its terminal velocity, $29$ m/s.  How long would it take to fall $381$ m?   (Hint: If we know a constant velocity and distance travelled, we can find the time.  Consider a similar question: if I have to drive 180 miles to Chicago, and drive 60 mph, how long will it take me?  How do I get that answer?)

Find the answer using Python tools below.

In [32]:
v = 29 #Units: m/s
s = 381 #Units: m
t = s/v
print(t)


13.137931034482758


---
<br>

🟨 🟨

### Exercise 3

We started with the question, "What is the velocity of the penny when it hits the ground?"  Our discovery that the terminal velocity is 29 m/s (65 mph) seems to answer that question: the velocity will be 29 m/s!  

<br>

Why is that a problematic answer?  Why do the *Mythbusters* guys say the terminal velocity is "between 35 and 65 mph" (and [this site](https://mechse.illinois.edu/news/blogs/lethal-pennies-and-other-misconceptions) gives a slightly different answer)?  What might affect the actual velocity?

✅ ✅ The model leaves out lot of important factors such as air resistance.



---
<br>

🟨 🟨

### Exercise 4

So far we have considered two models of a falling penny:

* If we ignore air resistance, the penny falls with constant acceleration, and we can compute the time to reach the sidewalk and the velocity of the penny when it gets there.

* If we take air resistance into account, and drop the penny at its terminal velocity, it falls with constant velocity.

Now let's consider a third model that includes elements of the first two: let's assume that the acceleration of the penny is `a` until the penny reaches $29$ m/s, and then $0$ m/s afterwards.  What is the total time for the penny to fall $381$ m?

This is a two phase simulation, because we have to break the problem into 2 time periods: when it is accelerating downward, and when its acceleration is zero.  To solve, answer the following 3 questions:

1. How long would the penny take to reach $29$ m/s with constant acceleration `g`.  When an object is accelerating at $9.8  m/s^2$, that means every second its velocity increases by $9.8  m/s$.
2. How far would it fall during that time?
3. How long would it take to fall the remaining distance with constant velocity $29$ m/s?

Suggestion: Assign each intermediate result to a variable with a meaningful name.   Here are the kinematic equations from the start of this notebook:

$$v = g t$$

$$x = g t^2 / 2$$

$$ v = \sqrt{2gx}$$

In [33]:
g = 9.8 #Units: m/s^2
v = 29 #Units: m/s
t = v/g
print(t)


2.9591836734693877


In [ ]:
# Find how long it will take the penny to reach terminal velocity.
# Assign that value to a new variable.


In [ ]:
# Find how far it will fall while it is reaching terminal velocity.
# Use kinematic equations.


In [ ]:
# Find how long it will take to fall the remaining distance at constant velocity


In [ ]:
# Determine the total time.
# Assign this to a variable, and output the value of this variable.


According to this model, it takes almost 15 seconds to fall to the ground, which is more than our first model (8.8 s) as well as our second model (13.1 s).